includes/class-magic-link.php

```diff
@@ -0,0 +1,801 @@
+	 *   @type string $token  The token.
+	 *   @type string $client Client hash.
+	 *   @type string $time   Token creation time.
+	 * }
+	 */
+	public static function validate_token( $user_id, $client, $token ) {
+		$errors = new \WP_Error();
+		$user   = \get_user_by( 'id', $user_id );
+
+		if ( ! $user ) {
+			$errors->add( 'invalid_user', __( 'User not found.', 'newspack' ) );
+		} elseif ( ! self::can_magic_link( $user->ID ) ) {
+			$errors->add( 'invalid_user_type', __( 'Not allowed for this user', 'newspack' ) );
+		} else {
+			$tokens = \get_user_meta( $user->ID, self::TOKENS_META, true );
+			if ( empty( $tokens ) || empty( $token ) ) {
+				$errors->add( 'invalid_token', __( 'Invalid token.', 'newspack' ) );
+			}
+		}
+
+		$valid_token = false;
+
+		if ( ! $errors->has_errors() ) {
+			$expire = time() - self::get_token_expiration_period();
+
+			foreach ( $tokens as $index => $token_data ) {
+				if ( $token_data['time'] < $expire ) {
+					unset( $tokens[ $index ] );
+
+				} elseif ( $token_data['token'] === $token ) {
+					$valid_token = $token_data;
+
+					/** If token data has a client hash, it must be equal to the user's. */
+					if ( ! empty( $token_data['client'] ) && $token_data['client'] !== $client ) {
+						$errors->add( 'invalid_client', __( 'Invalid client.', 'newspack' ) );
+					}
+
+					unset( $tokens[ $index ] );
+					break;
+				}
+			}
+
+			if ( empty( $valid_token ) ) {
+				$errors->add( 'expired_token', __( 'Token has expired.', 'newspack' ) );
+			}
+			self::clear_client_secret_cookie();
+
+			$tokens = array_values( $tokens );
+			\update_user_meta( $user->ID, self::TOKENS_META, $tokens );
+		}
+
+		return $errors->has_errors() ? $errors : $valid_token;
+	}
+
+	/**
+	 * Handle a reader authentication attempt using magic link token.
+	 *
+	 * @param int    $user_id User ID.
+	 * @param string $token   Token to authenticate.
+	 *
+	 * @return bool|\WP_Error Whether the user was authenticated or WP_Error.
+	 */
+	private static function authenticate( $user_id, $token ) {
+		if ( \is_user_logged_in() ) {
+			return false;
+		}
+
+		$user = \get_user_by( 'id', $user_id );
+
+		if ( ! $user ) {
+			return new \WP_Error( 'invalid_user', __( 'User not found.', 'newspack' ) );
+		}
+
+		$client     = self::get_client_hash( $user );
+		$token_data = self::validate_token( $user_id, $client, $token );
+
+		if ( \is_wp_error( $token_data ) ) {
+			return $token_data;
+		}
+
+		if ( empty( $token_data ) ) {
+			return false;
+		}
+
+		Reader_Activation::set_reader_verified( $user );
+		Reader_Activation::set_current_reader( $user->ID );
+
+		/**
+		 * Fires after a reader has been authenticated via magic link.
+		 *
+		 * @param \WP_User $user User that has been authenticated.
+		 */
+		do_action( 'newspack_magic_link_authenticated', $user );
+
+		return true;
+	}
+
+	/**
+	 * Process magic link token from request.
+	 */
+	public static function process_token_request() {
+		if ( ! Reader_Activation::is_enabled() ) {
```

Comment:
They could, but since these are all public methods I'd rather, in this case, have the check inside the method. I don't think the same for the `Reader_Activation` class because it's its own class/features.

We should also keep in mind that `Reader_Activation::is_enabled()` may still be used outside of the "experimental" flag. Reader activation could be interpreted as optional.